rfcs/00xx_UMA_Unified_Modular_Accelerator_Interface.md

```diff
@@ -0,0 +1,184 @@
+  * Custom primitives can be registered
+  * Lowering from Relay to S-TIR, using TOPI or custom primitives 
+  * Interface for registering accelerator-specific schedules and passes
+  * Execution of UMA schedules and passes on S-TIR
+  * Output:  NS-TIR(including tir.extern calls)
+  * UMALower baseclass (Python only) has to be inherited by accelerator-specific Lower classes (e.g. Accelerator A Lower, etc)
+* UMACodegen
+  * Input: NS-TIR(including tir.extern calls)
+  * Defaults to standard TVM codegen
+  * Intend is to provide a Python interface to insert/emit target code
+  * UMACodegen baseclass has to be inherited by accelerator-specific Codegen classes (e.g. Accelerator A Codegen, etc)
+  * Output: Target .c files
+
+The intention is to use TensorIR and Relax with MetaScheduler for optimization.
+
+
+Abbreviations:
+S-TIR: Schedulable TIR
+NS-TIR: Non-Schedulable TIR
+
+
+### File and class structure and Snippets as example for integration
+UMA provides a mostly python-based API. On the C++ side, new targets are registered using target hooks (RFC #0010). A generic `codegen.cc` handles the calls to the python side.
+```
+.
+├── codegen.cc
+└── targets.cc
+```
+```cpp
+TVM_REGISTER_TARGET_KIND("accelerator_A", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_A"))
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_A::TIRToRuntime);
+
+TVM_REGISTER_TARGET_KIND("accelerator_B", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_B"));
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_B::TIRToRuntime);
+```
+The python API is structured as shown below. The base class `UMABackend` functions as the core API. It uses the API points described in the [flow description](#flow-description) to register elements for the different stages.
+```
+.
+├── backend.py
+├── api
+│   ├── codegen.py
+│   ├── lower.py
+│   ├── partitioner.py
+│   └── utils.py
+├── accelerator_A
+│   ├── backend.py
+│   ├── codegen.py
+│   ├── passes.py
+│   ├── patterns.py
+│   ├── schedules.py
+│   └── strategies.py
+└── accelerator_B
+    └── ...
+```
+A custom accelerator is added by inheriting the `UMABackend`. New elements (e.g., passes, schedules) are added using a registration machanism.
+```python
+"""UMA backend for the UltraTrail accelerator"""
+
+class UltraTrailBackend(UMABackend):
+    def __init__(self):
+        super(UltraTrailBackend, self).__init__()
+
+        # Relay to Relay function registration
+        self._register_pattern("conv1d_relu", conv1d_relu_pattern())
+
+        self._register_relay_pass(1, ConfigGenerator())
+        self._register_relay_pass(2, BufferScopeAnnotator())
+
+        # Relay to TIR function registration
+        self._register_operator_strategy("nn.conv1d", custom_conv1d_strategy, plevel=9)
+
+        self._register_tir_schedule(insert_extern_calls)
```

Comment:
Can you explain what is the type of "insert_extern_calls" here ?

Reply:
Ack. So this is function that defines something like ? : 

```
# Get block by its name
block_b = sch.get_block("B")
# Get loops surronding the block
(i,) = sch.get_loops(block_b)
# Tile the loop nesting.
i_0, i_1, i_2 = sch.split(i, factors=[2, 2, 2])
```

I think a reference explanation how the IRModule get transformed using an example (a single pass for each point of registeration) of  can go a long way :).

Reply:
Yes, exactly.

> I think a reference explanation how the IRModule get transformed using an example (a single pass for each point of registeration) of can go a long way :).

I agree. We will add simple examples for each registration.

Reply:
With commit 0dda395eda42e7c286e06c503f6215e30bd1a2f8 we added a bit of [documentation](https://github.com/boschresearch/tvm-rfcs/blob/rfc_uma/rfcs/00xx_UMA_Unified_Modular_Accelerator_Interface.md#umabackend-references) for the API with small examples of what could be registered in each step.

Reply:
looks great to me! resolving...